spanner/read.go

```diff
@@ -646,7 +647,10 @@ func (d *resumableStreamDecoder) next(mt *builtinMetricsTracer) bool {
 			// to caller.
 			d.q.clear()
 			return false
-		case finished:
+		case finished, ended:
+			if d.state == ended {
+				cancel = nil
```

Comment:
Setting cancel to nil here does not have any effect. This only sets the local variable to nil. It does not stop `iterator.Stop` from calling `cancel()`, as that reference is not set to nil by this.

I just check this locally, and the current implementation has a random race condition:
1. `iterator.Stop` will always call `cancel()`, because the reference on the iterator is not set to nil.
2. The call to `cancel()` sometimes happen before the goroutine that fetches the trailers has executed (marking the stream as cancelled).
3. The call to `cancel()` can however also be executed after the goroutine that fetches the trailers has executed. Then the stream is marked as ended successfully.